bus-mapping/src/evm/opcodes/extcodesize.rs

```diff
@@ -12,34 +12,279 @@ impl Opcode for Extcodesize {
+                CallContextField::IsPersistent,
+                U256::from(state.call()?.is_persistent as u64),
+            ),
+        ] {
+            state.call_context_read(&mut exec_step, state.call()?.call_id, field, value);
+        }
+
+        // Update transaction access list for account address.
+        let is_warm = state.sdb.check_account_in_access_list(&address);
         state.push_op_reversible(
             &mut exec_step,
             RW::WRITE,
             TxAccessListAccountOp {
                 tx_id: state.tx_ctx.id(),
-                address: external_address,
+                address,
                 is_warm: true,
                 is_warm_prev: is_warm,
             },
         )?;
 
+        // Read account code hash and get code length.
+        let account = state.sdb.get_account(&address).1;
+        let exists = !account.is_empty();
+        let code_size = if exists {
+            let code_hash = account.code_hash;
+            let code_hash_word = code_hash.to_word();
+            state.account_read(
+                &mut exec_step,
+                address,
+                AccountField::CodeHash,
+                code_hash_word,
+                code_hash_word,
+            )?;
+            state.code(code_hash)?.len()
+        } else {
+            state.account_read(
+                &mut exec_step,
+                address,
+                AccountField::NonExisting,
+                Word::zero(),
+                Word::zero(),
+            )?;
+            0
+        };
+
+        // Write the EXTCODESIZE result to stack.
+        debug_assert_eq!(code_size, geth_steps[1].stack.last()?.as_usize());
         state.stack_write(
             &mut exec_step,
             geth_steps[1].stack.nth_last_filled(0),
-            geth_steps[1].stack.nth_last(0)?,
+            code_size.into(),
         )?;
 
         Ok(vec![exec_step])
     }
 }
+
+#[cfg(test)]
+mod extcodesize_tests {
+    use super::*;
+    use crate::circuit_input_builder::ExecState;
+    use crate::mock::BlockData;
+    use crate::operation::{AccountOp, CallContextOp, StackOp};
+    use eth_types::evm_types::{OpcodeId, StackAddress};
+    use eth_types::geth_types::{Account, GethData};
+    use eth_types::{address, bytecode, Address, Bytecode, Bytes, Word, U256};
+    use ethers_core::utils::keccak256;
+    use lazy_static::lazy_static;
+    use mock::TestContext;
+    use pretty_assertions::assert_eq;
+
+    lazy_static! {
+        static ref TEST_ADDRESS: Address = address!("0xaabbccddee000000000000000000000000000000");
+        static ref TEST_CODE: Bytes = Bytes::from([34, 54, 56]);
+        static ref TEST_ACCOUNT: Option<Account> = Some(Account {
+            address: *TEST_ADDRESS,
+            code: TEST_CODE.clone(),
+            ..Default::default()
+        });
+    }
+
+    #[test]
+    fn extcodesize_opcode_non_existing_account() {
+        test_ok(None, false);
+    }
+
+    #[test]
+    fn extcodesize_opcode_empty_account() {
```

Comment:
I think we can put these tests into one like: 
```
fn extcodesize_opcode_account() {
  // test empty account 
  test_ok(Some(Account::default()), false);
  // test cold account access
  test_ok(TEST_ACCOUNT.clone(), false);
  // test warm account access
   test_ok(TEST_ACCOUNT.clone(), true);
}

```

Reply:
Updated and merged test cases in bus-mapping as:
```
#[test]
fn test_extcodesize_opcode() {
    // Test for non existing account.
    test_ok(None, false);
    // Test for empty account.
    test_ok(Some(Account::default()), false);
    // Test for cold account.
    test_ok(TEST_ACCOUNT.clone(), false);
    // Test for warm account.
    test_ok(TEST_ACCOUNT.clone(), true);
}
```
And also merged the test cases in zkevm-circuit as:
```
    #[test]
    fn test_extcodesize_gadget() {
        // Test for non existing account.
        test_ok(None, false);
        // Test for empty account.
        test_ok(Some(Account::default()), false);
        // Test for cold account.
        test_ok(TEST_ACCOUNT.clone(), false);
        // Test for warm account.
        test_ok(TEST_ACCOUNT.clone(), true);
    }
```